// @livekit/agents-plugin-deepgram/dist/_utils.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var utils_exports = {};
__export(utils_exports, {
  PeriodicCollector: () => PeriodicCollector
});
module.exports = __toCommonJS(utils_exports);
class PeriodicCollector {
  duration;
  callback;
  lastFlushTime;
  total = null;
  constructor(callback, options) {
    this.duration = options.duration;
    this.callback = callback;
    this.lastFlushTime = performance.now() / 1e3;
  }
  push(value) {
    if (this.total === null) {
      this.total = value;
    } else {
      this.total = this.total + value;
    }
    if (performance.now() / 1e3 - this.lastFlushTime >= this.duration) {
      this.flush();
    }
  }
  flush() {
    if (this.total !== null) {
      this.callback(this.total);
      this.total = null;
    }
    this.lastFlushTime = performance.now() / 1e3;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PeriodicCollector
});
