// @livekit/agents-plugin-deepgram/dist/stt_v2.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var stt_v2_exports = {};
__export(stt_v2_exports, {
  STTv2: () => STTv2
});
module.exports = __toCommonJS(stt_v2_exports);
var import_agents = require("@livekit/agents");
var queryString = __toESM(require("node:querystring"), 1);
var import_ws = require("ws");
var import_utils = require("./_utils.cjs");
const _CLOSE_MSG = JSON.stringify({ type: "CloseStream" });
const defaultSTTv2Options = {
  model: "flux-general-en",
  sampleRate: 16e3,
  keyterms: [],
  endpointUrl: "wss://api.deepgram.com/v2/listen",
  language: "en",
  mipOptOut: false
};
function validateTags(tags) {
  for (const tag of tags) {
    if (tag.length > 128) {
      throw new Error("tag must be no more than 128 characters");
    }
  }
  return tags;
}
class STTv2 extends import_agents.stt.STT {
  label = "deepgram.STTv2";
  #opts;
  #apiKey;
  #logger = (0, import_agents.log)();
  /**
   * Create a new Deepgram STTv2 instance.
   *
   * @param opts - Configuration options
   * @param opts.apiKey - Deepgram API key (defaults to `DEEPGRAM_API_KEY` env var)
   * @param opts.model - Model to use (default: `flux-general-en`)
   * @param opts.eagerEotThreshold - Threshold (0.3-0.9) for preemptive generation
   * @param opts.eotThreshold - End-of-turn detection threshold (default: 0.7)
   * @param opts.eotTimeoutMs - End-of-turn timeout in ms (default: 3000)
   * @param opts.keyterms - List of key terms to improve recognition
   * @param opts.tags - Tags for usage reporting (max 128 chars each)
   *
   * @throws Error if no API key is provided
   */
  constructor(opts = {}) {
    super({
      streaming: true,
      interimResults: true,
      alignedTranscript: "word"
    });
    this.#opts = { ...defaultSTTv2Options, ...opts };
    const apiKey = opts.apiKey || process.env.DEEPGRAM_API_KEY;
    if (!apiKey) {
      throw new Error("Deepgram API key is required");
    }
    this.#apiKey = apiKey;
    if (this.#opts.tags) {
      this.#opts.tags = validateTags(this.#opts.tags);
    }
  }
  /** The model being used for transcription */
  get model() {
    return this.#opts.model;
  }
  /** The STT provider name */
  get provider() {
    return "Deepgram";
  }
  async _recognize(_frame, _abortSignal) {
    throw new Error("V2 API does not support non-streaming recognize. Use .stream()");
  }
  /**
   * Create a new streaming transcription session.
   *
   * @param options - Stream options
   * @returns A SpeechStream that emits transcription events
   */
  stream(options) {
    const streamOpts = { ...this.#opts, apiKey: this.#apiKey };
    return new SpeechStreamv2(this, streamOpts, options == null ? void 0 : options.connOptions);
  }
  /**
   * Update STT options. Changes will take effect on the next stream.
   *
   * @param opts - Partial options to update
   */
  updateOptions(opts) {
    this.#opts = { ...this.#opts, ...opts };
    if (opts.tags) this.#opts.tags = validateTags(opts.tags);
    this.#logger.debug("Updated STTv2 options");
  }
}
class SpeechStreamv2 extends import_agents.stt.SpeechStream {
  label = "deepgram.SpeechStreamv2";
  #opts;
  #logger = (0, import_agents.log)();
  #ws = null;
  #audioDurationCollector;
  #requestId = "";
  #speaking = false;
  // Parity: _reconnect_event - using existing Event class from @livekit/agents
  #reconnectEvent = new import_agents.Event();
  constructor(sttInstance, opts, connOptions) {
    super(sttInstance, opts.sampleRate, connOptions);
    this.#opts = opts;
    this.#audioDurationCollector = new import_utils.PeriodicCollector(
      (duration) => this.#onAudioDurationReport(duration),
      { duration: 5 }
    );
  }
  updateOptions(opts) {
    this.#logger.debug("Stream received option update", opts);
    this.#opts = { ...this.#opts, ...opts };
    if (opts.tags) this.#opts.tags = validateTags(opts.tags);
    this.#reconnectEvent.set();
  }
  async run() {
    var _a;
    while (!this.closed) {
      try {
        this.#reconnectEvent.clear();
        const url = this.#getDeepgramUrl();
        this.#logger.debug(`Connecting to Deepgram: ${url}`);
        this.#ws = new import_ws.WebSocket(url, {
          headers: { Authorization: `Token ${this.#opts.apiKey}` }
        });
        await new Promise((resolve, reject) => {
          if (!this.#ws) return reject(new Error("WebSocket not initialized"));
          const onOpen = () => {
            var _a2;
            (_a2 = this.#ws) == null ? void 0 : _a2.off("error", onError);
            resolve();
          };
          const onError = (err) => {
            var _a2;
            (_a2 = this.#ws) == null ? void 0 : _a2.off("open", onOpen);
            reject(err);
          };
          this.#ws.once("open", onOpen);
          this.#ws.once("error", onError);
        });
        const sendPromise = this.#sendTask();
        const recvPromise = this.#recvTask();
        const reconnectWait = this.#reconnectEvent.wait();
        const result = await Promise.race([
          Promise.all([sendPromise, recvPromise]),
          reconnectWait.then(() => "RECONNECT")
        ]);
        if (result === "RECONNECT") {
          this.#logger.debug("Reconnecting stream due to option update...");
          this.#ws.close();
        } else {
          break;
        }
      } catch (error) {
        this.#logger.error("Deepgram stream error", { error });
        throw error;
      } finally {
        if (((_a = this.#ws) == null ? void 0 : _a.readyState) === import_ws.WebSocket.OPEN) {
          this.#ws.close();
        }
      }
    }
    this.close();
  }
  async #sendTask() {
    if (!this.#ws) return;
    const samples50ms = Math.floor(this.#opts.sampleRate / 20);
    const audioBstream = new import_agents.AudioByteStream(this.#opts.sampleRate, 1, samples50ms);
    let hasEnded = false;
    const iterator = this.input[Symbol.asyncIterator]();
    while (true) {
      const nextPromise = iterator.next();
      const abortPromise = this.#reconnectEvent.wait().then(() => ({ abort: true }));
      const result = await Promise.race([nextPromise, abortPromise]);
      if ("abort" in result || result.done) {
        if (!("abort" in result) && result.done) {
          hasEnded = true;
        } else {
          break;
        }
      }
      if (hasEnded && !("value" in result)) {
      } else if ("value" in result) {
        const data = result.value;
        const frames = [];
        if (data === SpeechStreamv2.FLUSH_SENTINEL) {
          frames.push(...audioBstream.flush());
          hasEnded = true;
        } else {
          frames.push(...audioBstream.write(data.data.buffer));
        }
        for (const frame of frames) {
          this.#audioDurationCollector.push((0, import_agents.calculateAudioDurationSeconds)(frame));
          if (this.#ws.readyState === import_ws.WebSocket.OPEN) {
            this.#ws.send(frame.data);
          }
          if (hasEnded) {
            this.#audioDurationCollector.flush();
            hasEnded = false;
          }
        }
      }
      if (hasEnded) break;
    }
    if (!this.#reconnectEvent.isSet && this.#ws.readyState === import_ws.WebSocket.OPEN) {
      this.#logger.debug("Sending CloseStream message to Deepgram");
      this.#ws.send(_CLOSE_MSG);
    }
  }
  async #recvTask() {
    if (!this.#ws) return;
    return new Promise((resolve) => {
      if (!this.#ws) return resolve();
      this.#ws.on("message", (data, isBinary) => {
        if (isBinary) {
          this.#logger.warn("Received unexpected binary message from Deepgram");
          return;
        }
        try {
          const msg = JSON.parse(data.toString());
          this.#processStreamEvent(msg);
        } catch (error) {
          this.#logger.error("Failed to parse Deepgram message", { error });
        }
      });
      this.#ws.on("close", (code, reason) => {
        this.#logger.debug(`Deepgram WebSocket closed: ${code} ${reason}`);
        resolve();
      });
      this.#ws.on("error", () => resolve());
    });
  }
  #processStreamEvent(data) {
    if (data.request_id) {
      this.#requestId = data.request_id;
    }
    if (data.type === "TurnInfo") {
      const eventType = data.event;
      if (eventType === "StartOfTurn") {
        if (this.#speaking) return;
        this.#speaking = true;
        this.queue.put({
          type: import_agents.stt.SpeechEventType.START_OF_SPEECH,
          requestId: this.#requestId
        });
        this.#sendTranscriptEvent(import_agents.stt.SpeechEventType.INTERIM_TRANSCRIPT, data);
      } else if (eventType === "Update") {
        if (!this.#speaking) return;
        this.#sendTranscriptEvent(import_agents.stt.SpeechEventType.INTERIM_TRANSCRIPT, data);
      } else if (eventType === "EagerEndOfTurn") {
        if (!this.#speaking) return;
        this.#sendTranscriptEvent(import_agents.stt.SpeechEventType.PREFLIGHT_TRANSCRIPT, data);
      } else if (eventType === "TurnResumed") {
        this.#sendTranscriptEvent(import_agents.stt.SpeechEventType.INTERIM_TRANSCRIPT, data);
      } else if (eventType === "EndOfTurn") {
        if (!this.#speaking) return;
        this.#speaking = false;
        this.#sendTranscriptEvent(import_agents.stt.SpeechEventType.FINAL_TRANSCRIPT, data);
        this.queue.put({
          type: import_agents.stt.SpeechEventType.END_OF_SPEECH,
          requestId: this.#requestId
        });
      }
    } else if (data.type === "Error") {
      this.#logger.warn("deepgram sent an error", { data });
      const desc = data.description || "unknown error from deepgram";
      throw new Error(`Deepgram API Error: ${desc}`);
    }
  }
  #sendTranscriptEvent(eventType, data) {
    const alts = parseTranscription(this.#opts.language || "en", data, this.startTimeOffset);
    if (alts.length > 0) {
      this.queue.put({
        type: eventType,
        requestId: this.#requestId,
        alternatives: [alts[0], ...alts.slice(1)]
      });
    }
  }
  #onAudioDurationReport(duration) {
    const usageEvent = {
      type: import_agents.stt.SpeechEventType.RECOGNITION_USAGE,
      requestId: this.#requestId,
      recognitionUsage: {
        audioDuration: duration
      }
    };
    this.queue.put(usageEvent);
  }
  #getDeepgramUrl() {
    const params = {
      model: this.#opts.model,
      sample_rate: this.#opts.sampleRate.toString(),
      encoding: "linear16",
      mip_opt_out: String(this.#opts.mipOptOut)
    };
    if (this.#opts.eagerEotThreshold)
      params.eager_eot_threshold = this.#opts.eagerEotThreshold.toString();
    if (this.#opts.eotThreshold) params.eot_threshold = this.#opts.eotThreshold.toString();
    if (this.#opts.eotTimeoutMs) params.eot_timeout_ms = this.#opts.eotTimeoutMs.toString();
    if (this.#opts.keyterms.length > 0) params.keyterm = this.#opts.keyterms;
    if (this.#opts.tags && this.#opts.tags.length > 0) params.tag = this.#opts.tags;
    const baseUrl = this.#opts.endpointUrl.replace(/^http/, "ws");
    const qs = queryString.stringify(params);
    return `${baseUrl}?${qs}`;
  }
  close() {
    var _a;
    super.close();
    (_a = this.#ws) == null ? void 0 : _a.close();
  }
}
function parseTranscription(language, data, startTimeOffset) {
  const transcript = data.transcript;
  const wordsData = data.words || [];
  if (!wordsData || wordsData.length === 0) {
    return [];
  }
  let confidence = 0;
  if (wordsData.length > 0) {
    const sum = wordsData.reduce((acc, w) => acc + (w.confidence || 0), 0);
    confidence = sum / wordsData.length;
  }
  const sd = {
    language,
    startTime: (data.audio_window_start || 0) + startTimeOffset,
    endTime: (data.audio_window_end || 0) + startTimeOffset,
    confidence,
    text: transcript || "",
    // Note: Deepgram V2 (Flux) API does not provide word-level timing (start/end).
    // Words only contain 'word' and 'confidence' fields, so startTime/endTime will be 0.
    // See: https://developers.deepgram.com/docs/flux/nova-3-migration
    words: wordsData.map(
      (word) => (0, import_agents.createTimedString)({
        text: word.word ?? "",
        startTime: (word.start ?? 0) + startTimeOffset,
        endTime: (word.end ?? 0) + startTimeOffset,
        confidence: word.confidence ?? 0,
        startTimeOffset
      })
    )
  };
  return [sd];
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  STTv2
});
