// @livekit/agents-plugin-cartesia/dist/tts.js
import {
  APIConnectionError,
  APITimeoutError,
  AudioByteStream,
  Future,
  createTimedString,
  log,
  shortuuid,
  stream,
  tokenize,
  tts
} from "@livekit/agents";
import { request } from "node:https";
import { WebSocket } from "ws";
import {
  TTSDefaultVoiceId
} from "./models.js";
import {
  cartesiaMessageSchema,
  hasWordTimestamps,
  isChunkMessage,
  isDoneMessage,
  isErrorMessage
} from "./types.js";
const AUTHORIZATION_HEADER = "X-API-Key";
const VERSION_HEADER = "Cartesia-Version";
const VERSION = "2024-06-10";
const NUM_CHANNELS = 1;
const BUFFERED_WORDS_COUNT = 8;
const defaultTTSOptions = {
  model: "sonic-2",
  encoding: "pcm_s16le",
  sampleRate: 24e3,
  voice: TTSDefaultVoiceId,
  apiKey: process.env.CARTESIA_API_KEY,
  language: "en",
  baseUrl: "https://api.cartesia.ai",
  chunkTimeout: 5e3,
  wordTimestamps: true
};
class TTS extends tts.TTS {
  #opts;
  label = "cartesia.TTS";
  constructor(opts = {}) {
    const resolvedOpts = {
      ...defaultTTSOptions,
      ...opts
    };
    super(resolvedOpts.sampleRate || defaultTTSOptions.sampleRate, NUM_CHANNELS, {
      streaming: true,
      alignedTranscript: resolvedOpts.wordTimestamps ?? true
    });
    this.#opts = resolvedOpts;
    if (this.#opts.apiKey === void 0) {
      throw new Error(
        "Cartesia API key is required, whether as an argument or as $CARTESIA_API_KEY"
      );
    }
    if ((this.#opts.speed || this.#opts.emotion) && this.#opts.model !== "sonic-2-2025-03-07") {
      const logger = log();
      logger.warn(
        { model: this.#opts.model, speed: this.#opts.speed, emotion: this.#opts.emotion },
        "speed and emotion controls are only supported for model 'sonic-2-2025-03-07', see https://docs.cartesia.ai/developer-tools/changelog for details"
      );
    }
  }
  updateOptions(opts) {
    this.#opts = { ...this.#opts, ...opts };
    if ((this.#opts.speed || this.#opts.emotion) && this.#opts.model !== "sonic-2-2025-03-07") {
      const logger = log();
      logger.warn(
        { model: this.#opts.model, speed: this.#opts.speed, emotion: this.#opts.emotion },
        "speed and emotion controls are only supported for model 'sonic-2-2025-03-07', see https://docs.cartesia.ai/developer-tools/changelog for details"
      );
    }
  }
  synthesize(text, connOptions, abortSignal) {
    return new ChunkedStream(this, text, this.#opts, connOptions, abortSignal);
  }
  stream(options) {
    return new SynthesizeStream(this, this.#opts, options == null ? void 0 : options.connOptions);
  }
}
class ChunkedStream extends tts.ChunkedStream {
  label = "cartesia.ChunkedStream";
  #logger = log();
  #opts;
  #text;
  constructor(tts2, text, opts, connOptions, abortSignal) {
    super(text, tts2, connOptions, abortSignal);
    this.#text = text;
    this.#opts = opts;
  }
  async run() {
    const requestId = shortuuid();
    const bstream = new AudioByteStream(this.#opts.sampleRate, NUM_CHANNELS);
    const json = toCartesiaOptions(this.#opts);
    json.transcript = this.#text;
    const baseUrl = new URL(this.#opts.baseUrl);
    const doneFut = new Future();
    const req = request(
      {
        hostname: baseUrl.hostname,
        port: parseInt(baseUrl.port) || (baseUrl.protocol === "https:" ? 443 : 80),
        path: "/tts/bytes",
        method: "POST",
        headers: {
          [AUTHORIZATION_HEADER]: this.#opts.apiKey,
          [VERSION_HEADER]: VERSION
        },
        signal: this.abortSignal
      },
      (res) => {
        res.on("data", (chunk) => {
          for (const frame of bstream.write(chunk)) {
            this.queue.put({
              requestId,
              frame,
              final: false,
              segmentId: requestId
            });
          }
        });
        res.on("close", () => {
          for (const frame of bstream.flush()) {
            this.queue.put({
              requestId,
              frame,
              final: false,
              segmentId: requestId
            });
          }
          this.queue.close();
          if (!doneFut.done) doneFut.resolve();
        });
        res.on("error", (err) => {
          if (err.message === "aborted") return;
          this.#logger.error({ err }, "Cartesia TTS response error");
          if (!doneFut.done) doneFut.reject(err);
        });
      }
    );
    req.on("error", (err) => {
      if (err.name === "AbortError") return;
      this.#logger.error({ err }, "Cartesia TTS request error");
      if (!doneFut.done) doneFut.reject(err);
    });
    req.on("close", () => {
      if (!doneFut.done) doneFut.resolve();
    });
    req.write(JSON.stringify(json));
    req.end();
    try {
      await doneFut.await;
    } catch (e) {
      if (this.abortSignal.aborted) return;
      if (!this.queue.closed) this.queue.close();
      throw toRetryableConnectionError(e);
    }
  }
}
class SynthesizeStream extends tts.SynthesizeStream {
  #opts;
  #logger = log();
  #tokenizer = new tokenize.basic.SentenceTokenizer({
    minSentenceLength: BUFFERED_WORDS_COUNT
  }).stream();
  label = "cartesia.SynthesizeStream";
  constructor(tts2, opts, connOptions) {
    super(tts2, connOptions);
    this.#opts = opts;
  }
  updateOptions(opts) {
    this.#opts = { ...this.#opts, ...opts };
    if ((this.#opts.speed || this.#opts.emotion) && this.#opts.model !== "sonic-2-2025-03-07") {
      this.#logger.warn(
        { model: this.#opts.model, speed: this.#opts.speed, emotion: this.#opts.emotion },
        "speed and emotion controls are only supported for model 'sonic-2-2025-03-07', see https://docs.cartesia.ai/developer-tools/changelog for details"
      );
    }
  }
  async run() {
    const requestId = shortuuid();
    let closing = false;
    let sentenceStreamClosed = false;
    const sentenceStreamTask = async (ws2) => {
      const packet = toCartesiaOptions(this.#opts, true);
      for await (const event of this.#tokenizer) {
        const msg = {
          ...packet,
          context_id: requestId,
          transcript: event.token + " ",
          continue: true
        };
        ws2.send(JSON.stringify(msg));
      }
      const endMsg = {
        ...packet,
        context_id: requestId,
        transcript: " ",
        continue: false
      };
      ws2.send(JSON.stringify(endMsg));
      sentenceStreamClosed = true;
    };
    const inputTask = async () => {
      for await (const data of this.input) {
        if (data === SynthesizeStream.FLUSH_SENTINEL) {
          this.#tokenizer.flush();
          continue;
        }
        this.#tokenizer.pushText(data);
      }
      this.#tokenizer.endInput();
      this.#tokenizer.close();
    };
    const recvTask = async (ws2) => {
      const bstream = new AudioByteStream(this.#opts.sampleRate, NUM_CHANNELS);
      const eventChannel = stream.createStreamChannel();
      let lastFrame;
      let pendingTimedTranscripts = [];
      const sendLastFrame = (segmentId, final) => {
        if (lastFrame && !this.queue.closed) {
          this.queue.put({
            requestId,
            segmentId,
            frame: lastFrame,
            final,
            timedTranscripts: pendingTimedTranscripts.length > 0 ? pendingTimedTranscripts : void 0
          });
          lastFrame = void 0;
          pendingTimedTranscripts = [];
        }
      };
      let timeout = null;
      const clearTTSChunkTimeout = () => {
        if (timeout) {
          clearTimeout(timeout);
          timeout = null;
        }
      };
      const onMessage = (data) => {
        void eventChannel.write(data).catch((error) => {
          this.#logger.debug({ error }, "Failed writing Cartesia event to channel (likely closed)");
        });
      };
      const onClose = (code, reason) => {
        if (!closing) {
          this.#logger.debug(`WebSocket closed with code ${code}: ${reason.toString()}`);
        }
        clearTTSChunkTimeout();
        void eventChannel.close();
      };
      const onError = (err) => {
        this.#logger.error({ err }, "Cartesia WebSocket error");
        void eventChannel.close();
      };
      ws2.on("message", onMessage);
      ws2.on("close", onClose);
      ws2.on("error", onError);
      try {
        const reader = eventChannel.stream().getReader();
        while (!this.closed && !this.abortController.signal.aborted) {
          const result = await reader.read();
          if (result.done) break;
          const rawMsg = result.value;
          let serverMsg;
          try {
            const json = JSON.parse(rawMsg.toString());
            serverMsg = cartesiaMessageSchema.parse(json);
          } catch (parseErr) {
            this.#logger.warn({ parseErr }, "Failed to parse Cartesia message");
            continue;
          }
          if (isErrorMessage(serverMsg)) {
            this.#logger.error({ error: serverMsg.error }, "Cartesia returned error");
            continue;
          }
          const segmentId = serverMsg.context_id;
          if (this.#opts.wordTimestamps !== false && hasWordTimestamps(serverMsg)) {
            const wordTimestamps = serverMsg.word_timestamps;
            for (let i = 0; i < wordTimestamps.words.length; i++) {
              const word = wordTimestamps.words[i];
              const startTime = wordTimestamps.start[i];
              const endTime = wordTimestamps.end[i];
              if (word !== void 0 && startTime !== void 0 && endTime !== void 0) {
                pendingTimedTranscripts.push(
                  createTimedString({
                    text: word + " ",
                    // Add space after word for consistency
                    startTime,
                    endTime
                  })
                );
              }
            }
          }
          if (isChunkMessage(serverMsg)) {
            const audioBuffer = Buffer.from(serverMsg.data, "base64");
            const audioData = audioBuffer.buffer.slice(
              audioBuffer.byteOffset,
              audioBuffer.byteOffset + audioBuffer.byteLength
            );
            for (const frame of bstream.write(audioData)) {
              sendLastFrame(segmentId, false);
              lastFrame = frame;
            }
            clearTTSChunkTimeout();
            timeout = setTimeout(() => {
              this.#logger.debug(
                `Cartesia WebSocket TTS chunk stream timeout after ${this.#opts.chunkTimeout}ms`
              );
              ws2.close();
            }, this.#opts.chunkTimeout);
          } else if (isDoneMessage(serverMsg)) {
            if (sentenceStreamClosed) {
              for (const frame of bstream.flush()) {
                sendLastFrame(segmentId, false);
                lastFrame = frame;
              }
              sendLastFrame(segmentId, true);
              if (!this.queue.closed) {
                this.queue.put(SynthesizeStream.END_OF_STREAM);
              }
              if (segmentId === requestId) {
                closing = true;
                clearTTSChunkTimeout();
                ws2.close();
                break;
              }
            }
          }
        }
      } catch (err) {
        if (err instanceof Error && !err.message.includes("WebSocket closed")) {
          if (err.message.includes("Queue is closed") || err.message.includes("Channel is closed")) {
            this.#logger.warn(
              { err },
              "Channel closed during transcript processing (expected during disconnect)"
            );
          } else {
            this.#logger.error({ err }, "Error in recvTask from Cartesia WebSocket");
          }
        }
      } finally {
        ws2.off("message", onMessage);
        ws2.off("close", onClose);
        ws2.off("error", onError);
        clearTTSChunkTimeout();
      }
    };
    const wsUrl = this.#opts.baseUrl.replace(/^http/, "ws");
    const url = `${wsUrl}/tts/websocket?api_key=${this.#opts.apiKey}&cartesia_version=${VERSION}`;
    let ws;
    try {
      ws = await connectCartesiaWebSocket({
        url,
        timeoutMs: this.connOptions.timeoutMs,
        abortSignal: this.abortSignal
      });
      await Promise.all([inputTask(), sentenceStreamTask(ws), recvTask(ws)]);
    } catch (e) {
      if (this.abortSignal.aborted) {
        return;
      }
      throw toRetryableConnectionError(e);
    } finally {
      if (ws && ws.readyState !== WebSocket.CLOSED) {
        safeTerminateWebSocket(ws);
      }
    }
  }
}
const asError = (e) => e instanceof Error ? e : new Error(String(e));
const transientNetworkCodes = /* @__PURE__ */ new Set([
  "ETIMEDOUT",
  "ECONNRESET",
  "EAI_AGAIN",
  "ENETUNREACH",
  "ECONNREFUSED",
  "EHOSTUNREACH"
]);
const isRecord = (v) => {
  return v !== null && typeof v === "object";
};
const isAggregateErrorLike = (e) => {
  if (!isRecord(e)) return false;
  return e.name === "AggregateError" && Array.isArray(e.errors);
};
const hasErrorCode = (e, code) => {
  if (isRecord(e) && e.code === code) return true;
  if (isAggregateErrorLike(e)) {
    return e.errors.some((inner) => hasErrorCode(inner, code));
  }
  return false;
};
const hasAnyTransientCode = (e) => {
  if (isRecord(e) && typeof e.code === "string") {
    return transientNetworkCodes.has(e.code);
  }
  if (isAggregateErrorLike(e)) {
    return e.errors.some((inner) => hasAnyTransientCode(inner));
  }
  return false;
};
const toRetryableConnectionError = (e) => {
  const err = asError(e);
  const isTimeout = hasErrorCode(e, "ETIMEDOUT") || typeof err.message === "string" && err.message.includes("ETIMEDOUT");
  const message = isTimeout ? `Cartesia connection timed out` : `Cartesia connection failed: ${err.message || "unknown error"}`;
  return isTimeout ? new APITimeoutError({ message }) : new APIConnectionError({ message });
};
const waitForWsOpen = async ({
  ws,
  timeoutMs,
  abortSignal
}) => {
  if (abortSignal.aborted) {
    throw new Error("aborted");
  }
  const fut = new Future();
  let timeout;
  const cleanup = () => {
    if (timeout) clearTimeout(timeout);
    ws.off("open", onOpen);
    ws.off("error", onError);
    ws.off("close", onClose);
    abortSignal.removeEventListener("abort", onAbort);
  };
  const onOpen = () => fut.resolve();
  const onError = (err) => fut.reject(asError(err));
  const onClose = (code, reason) => fut.reject(
    new Error(`WebSocket closed before open (code=${code}, reason=${reason.toString()})`)
  );
  const onAbort = () => fut.reject(new Error("aborted"));
  ws.on("open", onOpen);
  ws.on("error", onError);
  ws.on("close", onClose);
  abortSignal.addEventListener("abort", onAbort, { once: true });
  if (timeoutMs > 0) {
    timeout = setTimeout(() => fut.reject(new Error("connect timeout")), timeoutMs);
  }
  try {
    await fut.await;
  } finally {
    cleanup();
  }
};
const safeTerminateWebSocket = (ws) => {
  try {
    ws.on("error", () => {
    });
  } catch {
  }
  try {
    if (ws.readyState === WebSocket.CONNECTING) {
      ws.close();
    } else {
      ws.terminate();
    }
  } catch {
  }
};
const connectCartesiaWebSocket = async ({
  url,
  timeoutMs,
  abortSignal
}) => {
  const connectOnce = async (family) => {
    const ws = new WebSocket(url, { handshakeTimeout: timeoutMs, family });
    try {
      await waitForWsOpen({ ws, timeoutMs, abortSignal });
      return ws;
    } catch (e) {
      safeTerminateWebSocket(ws);
      throw e;
    }
  };
  try {
    return await connectOnce();
  } catch (e) {
    if (hasAnyTransientCode(e) || isAggregateErrorLike(e)) {
      return await connectOnce(4);
    }
    throw e;
  }
};
const toCartesiaOptions = (opts, streaming = false) => {
  const voice = {};
  if (typeof opts.voice === "string") {
    voice.mode = "id";
    voice.id = opts.voice;
  } else {
    voice.mode = "embedding";
    voice.embedding = opts.voice;
  }
  const voiceControls = {};
  if (opts.speed) {
    voiceControls.speed = opts.speed;
  }
  if (opts.emotion) {
    voiceControls.emotion = opts.emotion;
  }
  if (Object.keys(voiceControls).length) {
    voice.__experimental_controls = voiceControls;
  }
  const result = {
    model_id: opts.model,
    voice,
    output_format: {
      container: "raw",
      encoding: opts.encoding,
      sample_rate: opts.sampleRate
    },
    language: opts.language,
    pronunciation_dict_id: opts.pronunciationDictId
  };
  if (streaming && opts.wordTimestamps !== false) {
    result.add_timestamps = true;
  }
  return result;
};
export {
  ChunkedStream,
  SynthesizeStream,
  TTS
};
