// @livekit/agents-plugin-cartesia/dist/types.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var types_exports = {};
__export(types_exports, {
  cartesiaChunkMessageSchema: () => cartesiaChunkMessageSchema,
  cartesiaDoneMessageSchema: () => cartesiaDoneMessageSchema,
  cartesiaErrorMessageSchema: () => cartesiaErrorMessageSchema,
  cartesiaFlushDoneMessageSchema: () => cartesiaFlushDoneMessageSchema,
  cartesiaMessageSchema: () => cartesiaMessageSchema,
  cartesiaServerMessageSchema: () => cartesiaServerMessageSchema,
  cartesiaTimestampsMessageSchema: () => cartesiaTimestampsMessageSchema,
  cartesiaWordTimestampsSchema: () => cartesiaWordTimestampsSchema,
  hasWordTimestamps: () => hasWordTimestamps,
  isChunkMessage: () => isChunkMessage,
  isDoneMessage: () => isDoneMessage,
  isErrorMessage: () => isErrorMessage,
  isFlushDoneMessage: () => isFlushDoneMessage,
  isTimestampsMessage: () => isTimestampsMessage
});
module.exports = __toCommonJS(types_exports);
var import_zod = require("zod");
const cartesiaWordTimestampsSchema = import_zod.z.object({
  words: import_zod.z.array(import_zod.z.string()),
  start: import_zod.z.array(import_zod.z.number()),
  end: import_zod.z.array(import_zod.z.number())
});
const cartesiaChunkMessageSchema = import_zod.z.object({
  type: import_zod.z.literal("chunk"),
  data: import_zod.z.string(),
  done: import_zod.z.boolean(),
  status_code: import_zod.z.number(),
  step_time: import_zod.z.number(),
  context_id: import_zod.z.string()
});
const cartesiaTimestampsMessageSchema = import_zod.z.object({
  type: import_zod.z.literal("timestamps"),
  done: import_zod.z.boolean(),
  status_code: import_zod.z.number(),
  context_id: import_zod.z.string(),
  word_timestamps: cartesiaWordTimestampsSchema
});
const cartesiaDoneMessageSchema = import_zod.z.object({
  type: import_zod.z.literal("done"),
  done: import_zod.z.boolean(),
  status_code: import_zod.z.number(),
  context_id: import_zod.z.string()
});
const cartesiaFlushDoneMessageSchema = import_zod.z.object({
  type: import_zod.z.literal("flush_done"),
  done: import_zod.z.boolean(),
  flush_done: import_zod.z.boolean(),
  flush_id: import_zod.z.number(),
  status_code: import_zod.z.number(),
  context_id: import_zod.z.string()
});
const cartesiaErrorMessageSchema = import_zod.z.object({
  type: import_zod.z.string(),
  done: import_zod.z.boolean(),
  error: import_zod.z.string(),
  status_code: import_zod.z.number(),
  context_id: import_zod.z.string()
});
const cartesiaServerMessageSchema = import_zod.z.discriminatedUnion("type", [
  cartesiaChunkMessageSchema,
  cartesiaTimestampsMessageSchema,
  cartesiaDoneMessageSchema,
  cartesiaFlushDoneMessageSchema
]);
const cartesiaMessageSchema = import_zod.z.union([
  cartesiaServerMessageSchema,
  cartesiaErrorMessageSchema
]);
function isChunkMessage(msg) {
  return "type" in msg && msg.type === "chunk";
}
function isTimestampsMessage(msg) {
  return "type" in msg && msg.type === "timestamps";
}
function isDoneMessage(msg) {
  return "type" in msg && msg.type === "done";
}
function isFlushDoneMessage(msg) {
  return "type" in msg && msg.type === "flush_done";
}
function isErrorMessage(msg) {
  return "error" in msg && typeof msg.error === "string";
}
function hasWordTimestamps(msg) {
  return isTimestampsMessage(msg);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  cartesiaChunkMessageSchema,
  cartesiaDoneMessageSchema,
  cartesiaErrorMessageSchema,
  cartesiaFlushDoneMessageSchema,
  cartesiaMessageSchema,
  cartesiaServerMessageSchema,
  cartesiaTimestampsMessageSchema,
  cartesiaWordTimestampsSchema,
  hasWordTimestamps,
  isChunkMessage,
  isDoneMessage,
  isErrorMessage,
  isFlushDoneMessage,
  isTimestampsMessage
});
