// @livekit/agents-plugin-cartesia/dist/types.js
import { z } from "zod";
const cartesiaWordTimestampsSchema = z.object({
  words: z.array(z.string()),
  start: z.array(z.number()),
  end: z.array(z.number())
});
const cartesiaChunkMessageSchema = z.object({
  type: z.literal("chunk"),
  data: z.string(),
  done: z.boolean(),
  status_code: z.number(),
  step_time: z.number(),
  context_id: z.string()
});
const cartesiaTimestampsMessageSchema = z.object({
  type: z.literal("timestamps"),
  done: z.boolean(),
  status_code: z.number(),
  context_id: z.string(),
  word_timestamps: cartesiaWordTimestampsSchema
});
const cartesiaDoneMessageSchema = z.object({
  type: z.literal("done"),
  done: z.boolean(),
  status_code: z.number(),
  context_id: z.string()
});
const cartesiaFlushDoneMessageSchema = z.object({
  type: z.literal("flush_done"),
  done: z.boolean(),
  flush_done: z.boolean(),
  flush_id: z.number(),
  status_code: z.number(),
  context_id: z.string()
});
const cartesiaErrorMessageSchema = z.object({
  type: z.string(),
  done: z.boolean(),
  error: z.string(),
  status_code: z.number(),
  context_id: z.string()
});
const cartesiaServerMessageSchema = z.discriminatedUnion("type", [
  cartesiaChunkMessageSchema,
  cartesiaTimestampsMessageSchema,
  cartesiaDoneMessageSchema,
  cartesiaFlushDoneMessageSchema
]);
const cartesiaMessageSchema = z.union([
  cartesiaServerMessageSchema,
  cartesiaErrorMessageSchema
]);
function isChunkMessage(msg) {
  return "type" in msg && msg.type === "chunk";
}
function isTimestampsMessage(msg) {
  return "type" in msg && msg.type === "timestamps";
}
function isDoneMessage(msg) {
  return "type" in msg && msg.type === "done";
}
function isFlushDoneMessage(msg) {
  return "type" in msg && msg.type === "flush_done";
}
function isErrorMessage(msg) {
  return "error" in msg && typeof msg.error === "string";
}
function hasWordTimestamps(msg) {
  return isTimestampsMessage(msg);
}
export {
  cartesiaChunkMessageSchema,
  cartesiaDoneMessageSchema,
  cartesiaErrorMessageSchema,
  cartesiaFlushDoneMessageSchema,
  cartesiaMessageSchema,
  cartesiaServerMessageSchema,
  cartesiaTimestampsMessageSchema,
  cartesiaWordTimestampsSchema,
  hasWordTimestamps,
  isChunkMessage,
  isDoneMessage,
  isErrorMessage,
  isFlushDoneMessage,
  isTimestampsMessage
};
