// @livekit/agents-plugin-openai/dist/responses/llm.js
import {
  APIConnectionError,
  APIStatusError,
  APITimeoutError,
  DEFAULT_API_CONNECT_OPTIONS,
  llm,
  toError
} from "@livekit/agents";
import OpenAI from "openai";
const defaultLLMOptions = {
  model: "gpt-4.1",
  apiKey: process.env.OPENAI_API_KEY,
  strictToolSchema: true
};
class LLM extends llm.LLM {
  #client;
  #opts;
  /**
   * Create a new instance of OpenAI Responses LLM.
   *
   * @remarks
   * `apiKey` must be set to your OpenAI API key, either using the argument or by setting the
   * `OPENAI_API_KEY` environment variable.
   */
  constructor(opts = defaultLLMOptions) {
    super();
    this.#opts = { ...defaultLLMOptions, ...opts };
    if (this.#opts.apiKey === void 0) {
      throw new Error("OpenAI API key is required, whether as an argument or as $OPENAI_API_KEY");
    }
    this.#client = this.#opts.client || new OpenAI({
      baseURL: this.#opts.baseURL,
      apiKey: this.#opts.apiKey
    });
  }
  label() {
    return "openai.responses.LLM";
  }
  get model() {
    return this.#opts.model;
  }
  chat({
    chatCtx,
    toolCtx,
    connOptions = DEFAULT_API_CONNECT_OPTIONS,
    parallelToolCalls,
    toolChoice,
    extraKwargs
  }) {
    const modelOptions = { ...extraKwargs || {} };
    parallelToolCalls = parallelToolCalls !== void 0 ? parallelToolCalls : this.#opts.parallelToolCalls;
    if (toolCtx && Object.keys(toolCtx).length > 0 && parallelToolCalls !== void 0) {
      modelOptions.parallel_tool_calls = parallelToolCalls;
    }
    toolChoice = toolChoice !== void 0 ? toolChoice : this.#opts.toolChoice;
    if (toolChoice) {
      modelOptions.tool_choice = toolChoice;
    }
    if (this.#opts.temperature !== void 0) {
      modelOptions.temperature = this.#opts.temperature;
    }
    if (this.#opts.store !== void 0) {
      modelOptions.store = this.#opts.store;
    }
    if (this.#opts.metadata) {
      modelOptions.metadata = this.#opts.metadata;
    }
    return new LLMStream(this, {
      model: this.#opts.model,
      client: this.#client,
      chatCtx,
      toolCtx,
      connOptions,
      modelOptions,
      strictToolSchema: this.#opts.strictToolSchema ?? true
    });
  }
}
class LLMStream extends llm.LLMStream {
  model;
  client;
  modelOptions;
  strictToolSchema;
  responseId;
  constructor(llm2, {
    model,
    client,
    chatCtx,
    toolCtx,
    connOptions,
    modelOptions,
    strictToolSchema
  }) {
    super(llm2, { chatCtx, toolCtx, connOptions });
    this.model = model;
    this.client = client;
    this.modelOptions = modelOptions;
    this.strictToolSchema = strictToolSchema;
    this.responseId = "";
  }
  async run() {
    let retryable = true;
    try {
      const messages = await this.chatCtx.toProviderFormat(
        "openai.responses"
      );
      const tools = this.toolCtx ? Object.entries(this.toolCtx).map(([name, func]) => {
        const oaiParams = {
          type: "function",
          name,
          description: func.description,
          parameters: llm.toJsonSchema(
            func.parameters,
            true,
            this.strictToolSchema
          )
        };
        if (this.strictToolSchema) {
          oaiParams.strict = true;
        }
        return oaiParams;
      }) : void 0;
      const requestOptions = { ...this.modelOptions };
      if (!tools) {
        delete requestOptions.tool_choice;
      }
      const stream = await this.client.responses.create(
        {
          model: this.model,
          input: messages,
          tools,
          stream: true,
          ...requestOptions
        },
        {
          timeout: this.connOptions.timeoutMs
        }
      );
      for await (const event of stream) {
        retryable = false;
        let chunk;
        switch (event.type) {
          case "error":
            this.handleError(event);
            break;
          case "response.created":
            this.handleResponseCreated(event);
            break;
          case "response.output_item.done":
            chunk = this.handleResponseOutputItemDone(event);
            break;
          case "response.output_text.delta":
            chunk = this.handleResponseOutputTextDelta(event);
            break;
          case "response.completed":
            chunk = this.handleResponseCompleted(event);
            break;
        }
        if (chunk) {
          this.queue.put(chunk);
        }
      }
    } catch (error) {
      if (error instanceof APIStatusError || error instanceof APITimeoutError || error instanceof APIConnectionError) {
        throw error;
      } else if (error instanceof OpenAI.APIConnectionTimeoutError) {
        throw new APITimeoutError({ options: { retryable } });
      } else if (error instanceof OpenAI.APIError) {
        throw new APIStatusError({
          message: error.message,
          options: {
            statusCode: error.status,
            body: error.error,
            requestId: error.requestID,
            retryable
          }
        });
      } else {
        throw new APIConnectionError({
          message: toError(error).message,
          options: { retryable }
        });
      }
    }
  }
  handleError(event) {
    throw new APIStatusError({
      message: event.message,
      options: {
        statusCode: -1,
        retryable: false
      }
    });
  }
  handleResponseCreated(event) {
    this.responseId = event.response.id;
  }
  handleResponseOutputItemDone(event) {
    let chunk;
    if (event.item.type === "function_call") {
      chunk = {
        id: this.responseId,
        delta: {
          role: "assistant",
          content: void 0,
          toolCalls: [
            llm.FunctionCall.create({
              callId: event.item.call_id || "",
              name: event.item.name,
              args: event.item.arguments
            })
          ]
        }
      };
    }
    return chunk;
  }
  handleResponseOutputTextDelta(event) {
    return {
      id: this.responseId,
      delta: {
        role: "assistant",
        content: event.delta
      }
    };
  }
  handleResponseCompleted(event) {
    if (event.response.usage) {
      return {
        id: this.responseId,
        usage: {
          completionTokens: event.response.usage.output_tokens,
          promptTokens: event.response.usage.input_tokens,
          promptCachedTokens: event.response.usage.input_tokens_details.cached_tokens,
          totalTokens: event.response.usage.total_tokens
        }
      };
    }
    return void 0;
  }
}
export {
  LLM,
  LLMStream
};
